back-end/routes/clusters.js

```diff
@@ -0,0 +1,208 @@
+        }
+    });
+    return token ? token.token : null;
+};
+
+function getValenceFeatures(audioFeatures) {
+    const valenceFeatures = [];
+    audioFeatures.forEach(feature => {
+        const songId = feature.uri.split(':').pop();
+        const valenceValue = feature.valence;
+        valenceFeatures.push({ id: songId, valence: valenceValue });
+    });
+    return valenceFeatures;
+}
+
+async function fetchAndProcessAudioFeatures(songs, accessToken) {
+    const features = await fetchAudioFeatures(songs, accessToken);
+    const valenceFeatures = getValenceFeatures(features);
+    return valenceFeatures;
+}
+
+function euclideanDistance(a, b) {
+    return Math.abs(a - b);
+}
+
+// Calculates mean of valence value of a cluster
+function calculateMean(cluster) {
+    let sum = 0;
+    for (let i = 0; i < cluster.length; i++) {
+        sum += cluster[i].valence;
+    }
+    return sum / cluster.length;
+}
+
+// Initializes first centroid from one random song and then for the rest find the largest distance between a song and existing centroid
+// in order to obtain better initial centroids (K++ means clustering)# of Centroids = # of Clusters
+function initializeCentroids(songs, numClusters) {
+    let centroids = [];
+
+    centroids.push(songs[Math.floor(Math.random() * songs.length)].valence);
+
+    for (let i = 1; i < numClusters; i++) {
+        let distances = songs.map(song => {
+            let minDistance = Math.min(...centroids.map(centroid =>
+                Math.pow(song.valence - centroid, 2)));
+            return minDistance;
+        });
+
+        let totalDistance = 0;
+        for (let j = 0; j < distances.length; j++) {
+            totalDistance += distances[j]
+        }
+
+        let randomValue = Math.random() * totalDistance;
+        let runningSum = 0;
+
+        for (let j = 0; j < distances.length; j++) {
+            runningSum += distances[j];
+            if (runningSum > randomValue) {
+                centroids.push(songs[j].valence);
+                break;
+            }
+        }
+    }
+
+    return centroids;
+}
+
+// Finds the distance between each song's valence value and the valence value of the centroids using Euclidean formula then associates song with closest centroid
+function assignSongsToCentroids(songs, centroids) {
+    return songs.map(song => {
+        const distances = centroids.map(centroid => euclideanDistance(song.valence, centroid));
+        const closestCentroidIndex = distances.indexOf(Math.min(...distances));
+        return { id: song.id, valence: song.valence, cluster: closestCentroidIndex };
+    });
+}
+
+// Finds new centroid based on mean valence value of songs in the cluster
+function updateCentroids(assignments, numClusters) {
+    const newCentroids = [];
+    for (let i = 0; i < numClusters; i++) {
+        const clusterSongs = assignments.filter(song => song.cluster === i);
+        newCentroids[i] = calculateMean(clusterSongs);
+    }
+    return newCentroids;
+}
```

Comment:
So this is only happening for the valence, right? Can we have the functions' name reflect that?

Reply:
Actually, we're running the clustering for valence as a whole, right? We won't do it for any other feature at the moment, so we might want to rename the file as a whole ('clustersValence')?

Or any other idea at the top of your mind to be clear that the algorithm is valence specific? We can probably have two files, one which contains functions that can be reused for clustering as a whole, and another for valence specifically. And we might want to think on making our functions more general so they can be reused for other features and are not limited to valence.